core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -237,6 +266,29 @@ public void setTransportSupplier(final Supplier<ClientTransport> supplier) {
     }
   }
 
+  /**
+   * Sets all pending streams with a fail fast call option of the delayed transport as
+   * {@link FailingClientStream}s, and removes them from the list of pending streams of the
+   * transport.
+   * @param status the causal status for the clearing.
+   */
+  void clearFailFastPendingStreams(Status status) {
+    synchronized (lock) {
+      if (pendingStreams != null && !pendingStreams.isEmpty()) {
+        final Collection<PendingStream> savedPendingStreams = pendingStreams;
+        Iterator<PendingStream> it = savedPendingStreams.iterator();
+        while (it.hasNext()) {
+          PendingStream stream = it.next();
+          if (stream.isFailFast()) {
+            stream.setStream(new FailingClientStream(Status.UNAVAILABLE.withDescription(
```

Comment:
We really want this to be done within `streamCreationExecutor.execute()` to avoid deadlocks and starving an important thread. You should probably loop over all the `pendingStreams` and move fail fast streams to a new List. The new list would then be used on within a `Runnable` passed to `streamCreationExecutor`.